### PR TITLE
!htc #20371 allows for simple HTTPS usage from Java, fixes typo in APIs

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/20371
Please have a look @drewhk 

This is still the old DSL, but this should be backported to 2.0 if possible, since we want to allow https there too in a simple way.

The test will fail as it won't trust the certificate (which is good)

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 1]
```diff
@@ -764,6 +765,13 @@ trait DefaultSSLContextCreation {
   def createDefaultClientHttpsContext(): HttpsConnectionContext =
     createClientHttpsContext(sslConfig)
 
+  // currently the same configuration as client by default, however we should tune this for server-side apropriately (!)
+  def createServerHttpsContext(sslConfig: AkkaSSLConfig): HttpsConnectionContext = {
+    log.warning("Automatic server-side configuration is not supported yet, will attempt to use client-side settings. " +
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: Instead as part of this PR I'll start doing the "simple server side using ssl-config".

**Comment by ktoso** [comment]
TODO here:
- [ ] made a simple to follow docs page about this
- [ ] the defaultServerContext can be taken via ssl-config already – we'll just want the locateion of keystore and pass (pass would be injected via env variable)

**Comment by ktoso** [comment]
Please review @akka/akka-team

**Comment by ktoso** [comment]
Refers to http://typesafehub.github.io/ssl-config/CertificateGeneration.html

File: akka-http-core/src/main/scala/akka/http/javadsl/Http.scala [hunk 2]
```diff
@@ -640,14 +640,20 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
...
+   * Can be modified using [[setDefaultServerHttpContext]], and will then apply for servers bound after that call has completed.
    */
   def defaultServerHttpContext: ConnectionContext =
     delegate.defaultServerHttpContext
 
   /**
+   * Sets the default server-side [[ConnectionContext]].
+   * If it is an instance of [[HttpsConnectionContext]] then the server will be bound using HTTPS.
+   */
+  def setDefaultServerHttpContext(context: ConnectionContext): Unit =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: [FIX] this method was missing

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 3]
```diff
@@ -520,7 +521,7 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
    * Sets the default server-side [[ConnectionContext]].
    * If it is an instance of [[HttpsConnectionContext]] then the server will be bound using HTTPS.
    */
-  def setDefaultClientHttpsContext(context: ConnectionContext): Unit =
+  def setDefaultServerHttpContext(context: ConnectionContext): Unit =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: [FIX] this method had a typo

File: akka-docs/rst/scala/http/low-level-server-side-api.rst [hunk 4]
```diff
@@ -239,7 +210,7 @@ Bind failures
 
 The first type of failure is when the server is unable to bind to the given port. For example when the port
 is already taken by another application, or if the port is privileged (i.e. only usable by ``root``).
-In this case the "binding future" will fail immediatly, and we can react to if by listening on the Future's completion:
+In this case the "binding future" will fail immediatly, and we can react to if by listeg on the Future's completion:
```
Comment (johanandren) [inline akka-docs/rst/scala/http/low-level-server-side-api.rst]: listening

File: akka-docs/rst/scala/http/server-side-https-support.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+front of your application`_ instead of terminating TLS inside the JVM, and therefore Akka HTTP, itself.
+
+Other excellent articles on the subject:
+
+- `Oracle Java SE 8: Creating a Keystore using JSSE <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html#CreateKeystore>`_
+- `Java PKI Programmer's Guide <https://docs.oracle.com/javase/8/docs/technotes/guides/security/certpath/CertPathProgGuide.html>`_
+- `Fixing X.509 Certificates <https://tersesystems.com/2014/03/20/fixing-x509-certificates/>`_
+
+.. _Play documentation about securing your app: https://www.playframework.com/documentation/2.5.x/ConfiguringHttps#ssl-certificates
+.. _reverse proxy to terminate TLS in front of your application: https://www.playframework.com/documentation/2.5.x/HTTPServer
```
Comment (johanandren) [inline akka-docs/rst/scala/http/server-side-https-support.rst]: great with this entire section 👍

File: project/MiMa.scala [hunk 6]
```diff
@@ -744,6 +744,10 @@ object MiMa extends AutoPlugin {
         ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.recoverWithRetries")
       ),
       "2.4.4" -> Seq(
+        // #20371, missing method and typo in another one making it impossible to use HTTPs via setting default HttpsConnectionContext
+        ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.http.scaladsl.HttpExt.setDefaultClientHttpsContext"),
```
Comment (johanandren) [inline project/MiMa.scala]: this isn't a problem BC-wise?

**Comment by johanandren** [comment]
LGTM (with one spelling and one BC question as answered/fixed)

File: project/MiMa.scala (hunk 6 above)
Comment (ktoso) [inline project/MiMa.scala]: The meth may have unintended side effects... yeah. – oh wait, method! ;-)

So the story here is that it was completely broken for everyone - if you called this method you likely did not get what you wanted.

**Pull request merged**